runtime/permissions/lib.rs

```diff
@@ -324,27 +325,46 @@ pub trait Descriptor: Eq + Clone + Hash {
     api_name: Option<&str>,
   ) -> Result<(), AnyError>;
 
-  fn flag_name() -> &'static str;
-  fn name(&self) -> Cow<str>;
-  // By default, specifies no-stronger-than relationship.
-  // As this is not strict, it's only true when descriptors are the same.
-  fn stronger_than(&self, other: &Self) -> bool {
-    self == other
+  fn matches_allow(&self, other: &Self::AllowDesc) -> bool;
+  fn matches_deny(&self, other: &Self::DenyDesc) -> bool;
+
+  /// Gets if this query descriptor should revoke the provided allow descriptor.
+  fn revokes(&self, other: &Self::AllowDesc) -> bool;
+  fn stronger_than_deny(&self, other: &Self::DenyDesc) -> bool;
+}
+
+fn format_display_name(display_name: Cow<str>) -> String {
+  if display_name.starts_with('<') && display_name.ends_with('>') {
+    display_name.into_owned()
+  } else {
+    format!("\"{}\"", display_name)
   }
 }
 
-#[derive(Clone, Debug, Eq, PartialEq)]
-pub struct UnaryPermission<T: Descriptor + Hash> {
+pub trait AllowDescriptor: Debug + Eq + Clone + Hash {
+  type Query: QueryDescriptor;
+
+  fn as_query(&self) -> Self::Query;
+}
+
+pub trait DenyDescriptor: Debug + Eq + Clone + Hash {
+  type Query: QueryDescriptor;
+
+  fn is_partial(&self, query: &Self::Query) -> bool;
```

Comment:
I'll look into shifting this one over to query as well (ran out of battery where I am)